src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+                bytesReceived = received;
+                errorCode = SocketError.Success;
+                return true;
+            }
+
+            bytesReceived = 0;
+
+            if (errno != Interop.Error.EAGAIN && errno != Interop.Error.EWOULDBLOCK)
+            {
+                errorCode = SocketPal.GetSocketErrorForErrorCode(errno);
+                return true;
+            }
+
+            errorCode = SocketError.Success;
+            return false;
+        }
+
+        private static unsafe int Send(int fd, int flags, byte[] buffer, ref int offset, ref int count, byte[] socketAddress, int socketAddressLen, out Interop.Error errno)
+        {
+            var pinnedSocketAddress = default(GCHandle);
+            Interop.libc.sockaddr* sockAddr = null;
+            uint sockAddrLen = 0;
+
+            int sent;
+            try
+            {
+                if (socketAddress != null)
+                {
+                    pinnedSocketAddress = GCHandle.Alloc(socketAddress, GCHandleType.Pinned);
+                    sockAddr = (Interop.libc.sockaddr*)pinnedSocketAddress.AddrOfPinnedObject();
+                    sockAddrLen = (uint)socketAddressLen;
+                }
+
+                fixed (byte* b = buffer)
+                {
+                    sent = (int)Interop.libc.sendto(fd, &b[offset], (IntPtr)count, flags, sockAddr, sockAddrLen);
+                }
+            }
+            finally
+            {
+                if (pinnedSocketAddress.IsAllocated)
+                {
+                    pinnedSocketAddress.Free();
+                }
+            }
+
+            if (sent == -1)
+            {
+                errno = Interop.Sys.GetLastError();
+                return -1;
+            }
+
+            errno = Interop.Error.SUCCESS;
+            offset += sent;
+            count -= sent;
+            return sent;
+        }
+
+        private static unsafe int Send(int fd, int flags, BufferList buffers, ref int bufferIndex, ref int offset, byte[] socketAddress, int socketAddressLen, out Interop.Error errno)
+        {
+            // Pin buffers and set up iovecs
+            int startIndex = bufferIndex, startOffset = offset;
+
+            var pinnedSocketAddress = default(GCHandle);
+            Interop.libc.sockaddr* sockAddr = null;
+            uint sockAddrLen = 0;
+
+            int maxBuffers = buffers.Count - startIndex;
+            var handles = new GCHandle[maxBuffers];
+            var iovecs = new Interop.libc.iovec[maxBuffers];
+
+            int sent;
+            int toSend = 0, iovCount = maxBuffers;
+            try
+            {
+                for (int i = 0; i < maxBuffers; i++, startOffset = 0)
+                {
+                    ArraySegment<byte> buffer = buffers[startIndex + i];
+                    handles[i] = GCHandle.Alloc(buffer.Array, GCHandleType.Pinned);
+                    iovecs[i].iov_base = &((byte*)handles[i].AddrOfPinnedObject())[buffer.Offset + startOffset];
```

Comment:
Is it possibl for the caller-provided offset to extend beyond the length of the first buffer?